src/serialization/serialization.rs

```diff
@@ -0,0 +1,412 @@
+        self.buf.write_u32::<BigEndian>(h.significant_value_digits as u32)?;
+        self.buf.write_u64::<BigEndian>(h.lowest_discernible_value)?;
+        self.buf.write_u64::<BigEndian>(h.highest_trackable_value)?;
+        // int to double conversion
+        self.buf.write_f64::<BigEndian>(1.0)?;
+
+        debug_assert_eq!(V2_HEADER_SIZE, self.buf.len());
+
+        unsafe {
+            // want to treat the rest of the vec as a slice, and we've already reserved this
+            // space, so this way we don't have to resize() on a lot of dummy bytes.
+            self.buf.set_len(max_size);
+        }
+
+        let counts_len = encode_counts(h, &mut self.buf[V2_HEADER_SIZE..])?;
+        let total_len = V2_HEADER_SIZE + counts_len;
+
+        // TODO benchmark fastest buffer management scheme
+        // counts is always under 2^24
+        (&mut self.buf[4..8]).write_u32::<BigEndian>(counts_len as u32)?;
+
+        writer.write_all(&mut self.buf[0..(total_len)])
+            .map(|_| total_len)
+            .map_err(|_| SerializeError::IoError)
+    }
+
+    fn max_encoded_size<T: Counter>(h: &Histogram<T>) -> Option<usize> {
+        Self::counts_array_max_encoded_size(h.index_for(h.max()) + 1)
+            .and_then(|x| x.checked_add(V2_HEADER_SIZE))
+    }
+
+    fn counts_array_max_encoded_size(length: usize) -> Option<usize> {
+        // LEB128-64b9B uses at most 9 bytes
+        // Won't overflow (except sometimes on 16 bit systems) because largest possible counts
+        // len is 47 buckets, each with 2^17 half count, for a total of 6e6. This product will
+        // therefore be about 5e7 (50 million) at most.
+        length.checked_mul(9)
+    }
+}
+
+/// Errors that can happen during deserialization.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum DeserializeError {
+    /// An i/o operation failed.
+    IoError,
+    /// The cookie (first 4 bytes) did not match that for any supported format.
+    InvalidCookie,
+    /// The histogram uses features that this implementation doesn't support (yet), so it cannot
+    /// be deserialized correctly.
+    UnsupportedFeature,
+    /// A count exceeded what can be represented in the chosen counter type.
+    UnsuitableCounterType,
+    /// The histogram instance could not be created because the serialized parameters were invalid
+    /// (e.g. lowest value, highest value, etc.)
+    InvalidParameters,
+    /// The current system's pointer width cannot represent the encoded histogram.
+    UsizeTypeTooSmall,
+    /// The encoded array is longer than it should be for the histogram's value range.
+    EncodedArrayTooLong
+}
+
+impl std::convert::From<std::io::Error> for DeserializeError {
+    fn from(_: std::io::Error) -> Self {
+        DeserializeError::IoError
+    }
+}
+
+/// Deserializer for all supported formats.
+/// Deserializers are intended to be re-used for many histograms.
+pub struct Deserializer {
+    payload_buf: Vec<u8>
+}
+
+impl Deserializer {
```

Comment:
I was thinking it would probably be nice to follow in the footsteps of the Java implementation, at least in spirit: there's one deserialize function and it knows how to unpack every flavor. Given the "cookie", it is easy to do the detection for the user.